[WIP] Fixes for 0.1

tarruda opened the pull request:
The goal of this PR is to fix bugs tagged "milestone:0.1-first-public-release"

tarruda committed:
ui_bridge: Fix passing NULL pointer to memcpy

tarruda committed:
os/path: Fix path_get_absolute_path for top-level paths

Close #2833

justinmk commented on src/nvim/ui_bridge.c:
```diff
@@ -234,8 +234,11 @@ static void ui_bridge_highlight_set_event(void **argv)
 
 static void ui_bridge_put(UI *b, uint8_t *text, size_t size)
```
would FUNC_ATTR_NONNULL_ALL work?

tarruda replied:
No, this function uses NULL in the second cell of double-width cells

ZyX-I replied:
Should not it be `FUNC_ATTR_NONNULL_ARG(1)` then?

tarruda replied:
> Should not it be FUNC_ATTR_NONNULL_ARG(1) then?

:+1:

justinmk commented on src/nvim/path.c:
```diff
@@ -2080,8 +2080,14 @@ static int path_get_absolute_path(const char_u *fname, char_u *buf, int len, int
-      STRNCPY(relative_directory, fname, p-fname);
-      relative_directory[p-fname] = NUL;
+      if (p == fname) {
+        // relative to root
```
In other words:

    // fname is an absolute path

tarruda replied:
I didn't explain well, this path is executed if there's only one component AND it's relative to root

justinmk replied:
Ok. Only purpose of my suggestion was to improve the wording ("relative to root" is an odd way to say "absolute").  The unit test implies that `force=true` is required, so `path_is_absolute_path(fname)` must be true.

tarruda committed:
regexp: Fix invalid multibyte reads at end of strings

Close #3150

tarruda committed:
eval: Fix `ex_delfunction` to use the `uf_refcount` field properly

@4b98ea1e80bf changed how refcounts are handled internally to fit into job
control semantics. Change the refcount check in `ex_delfunction` to consider
this. Close #3000

tarruda committed:
terminal: Don't enter terminal mode from `:normal`

Ref: https://github.com/junegunn/fzf.vim/issues/8#issuecomment-139209765

justinmk commented on src/nvim/search.c:
```diff
@@ -653,7 +653,7 @@ int searchit(
               if (matchcol == 0 && (options & SEARCH_START))
                 break;
-              if (ptr[matchcol] == NUL
+              if (STRLEN(ptr) <= (size_t)matchcol || ptr[matchcol] == NUL
```
The STRLEN calculation could be moved out of the while-loop (above lin 614).

tarruda replied:
Did you see the comment below(at `ml_get_buf`)? Even though it is theoretically getting the same line, is it guaranteed that `ptr` will point to the same string? I'd rather not optimize this prematurely and risk introducing bugs, especially because vim is too complex and hard to have its state predicted after a non-trivial function is called(`vim_regexec_multi`)

justinmk commented on src/nvim/regexp_nfa.c:
```diff
@@ -4781,6 +4781,7 @@ static long find_match_text(colnr_T startcol, int regstart, char_u *match_text)
     if (match
         /* check that no composing char follows */
         && !(enc_utf8
+             && STRLEN(regline) > (size_t)(col + len2)
```
The STRLEN calculation could be moved out of the for-loop.

tarruda replied:
Same as above, `regline` is a global variable and there are some function calls in the loop. Unless you inspected every possible code path in the block, it is safer to just leave it there.

justinmk commented:
https://github.com/neovim/neovim/issues/3016 is another one that 49a850c should fix.

tarruda committed:
ugrid: Increase cell size to allow unicode composition

Close #3323

tarruda committed:
terminal: Run screen refresh timer in a deferred queue

Close #3332

tarruda commented:
@oni-link continuing the conversation from a rebased commit, you asked:

> Groups of 'maxcombine' combining characters. What does vim do? Drop the remaining characters?

From the `'maxcombine'` help:

> 	The default is OK for most languages.  Hebrew may require 4.
	Maximum value is 6.

So I don't think it's possible to have more than 6, which is confirmed by the size of the global variable which stores combining chars:

```c
EXTERN u8char_T *ScreenLinesC[MAX_MCO];         /* composing characters */
```

tarruda committed:
terminal: Don't store a buf_T reference in the terminal struct

Since vimscript can close buffers at any time, it is possible that a
refresh_timer_cb will be called with an invalid buffer, but there's no way to
detect this if only a reference is stored because the memory can be reused by
the allocator. Use buf_T->handle which is guaranteed to be unique.

Tranquility commented:
This PR fixes https://github.com/junegunn/fzf/issues/206 for me.

tarruda commented:
@Tranquility great

Since this PR already fixes quite a few issues, I will merge it later and continue working on the remaining 0.1 issues in a new PR

justinmk commented on src/nvim/ui_bridge.c:
```diff
@@ -233,8 +234,11 @@ static void ui_bridge_highlight_set_event(void **argv)
-  memcpy(t, text, size);
+  uint8_t *t = NULL;
+  if (text) {
+    t = xmalloc(sizeof(((UCell *)0)->data));
```
Interesting. http://stackoverflow.com/a/19786755/152142
